File: python3源码/Test25.py
# -*- coding:utf-8 -*-

# 剑指offer-题25：二叉树中和为某一值的路径

# 输入一颗二叉树和一个整数，打印出二叉树中结点值的和为输入整数的所有路径。路径定
# 义为从树的根结点开始往下一直到叶结点所经过的结点形成一条路径。


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    # 返回二维列表，内部每个列表表示找到的路径
    def FindPath(self, root, expectNumber):
        self.path_list = []
        self.stack = []
        self.cur_sum = 0
        return self.get_path(root, expectNumber)

    def get_path(self, root, expectNumber):
        if root is None:
            return self.path_list
        self.cur_sum += root.val
        self.stack.append(root.val)

        is_leaf = root.right is None and root.left is None
        if self.cur_sum == expectNumber and is_leaf:
            temp_path = self.stack[:]
            self.path_list.append(temp_path)
            # self.path_list.append(self.stack.copy())
        if root.left:
            self.get_path(root.left, expectNumber)
        if root.right:
            self.get_path(root.right, expectNumber)
        self.stack.pop()
        self.cur_sum -= root.val
        return self.path_list


solution = Solution()
root = TreeNode(10)
a = TreeNode(5)
b = TreeNode(4)
c = TreeNode(7)
d = TreeNode(12)
root.left = a
a.left = b
a.right = c
root.right = d
print(solution.FindPath(root, 22))
print(solution.FindPath(root, 15))

File: java源码/offer/Test25.java
package offer;

import java.util.ArrayList;
import java.util.Stack;

/**
 * 剑指offer-题25：二叉树中和为某一值的路径
 * 
 * 输入一颗二叉树和一个整数，打印出二叉树中结点值的和为输入整数的所有路径。路径定义为从树的根结点开始往下一直到叶结点所经过的结点形成一条路径。
 */

public class Test25 {
	int curSum = 0;
	Stack<Integer> pathStack = new Stack<>();
	ArrayList<ArrayList<Integer>> pathList = new ArrayList<>();

	public ArrayList<ArrayList<Integer>> FindPath(TreeNode root, int target) {
		if (root == null) {
			return pathList;
		}
		curSum += root.val;
		pathStack.push(root.val);

		boolean isLeafNode = root.left == null && root.right == null;
		if (curSum == target && isLeafNode) {
			ArrayList<Integer> path = new ArrayList<Integer>();
			for (Integer value : pathStack) {
				path.add(value);
			}
			pathList.add(path);
		}
		if (root.left != null) {
			FindPath(root.left, target);
		}
		if (root.right != null) {
			FindPath(root.right, target);
		}
		pathStack.pop();
		curSum -= root.val;
		return pathList;
	}

	public static void main(String[] args) {
		TreeNode root = new TreeNode(10);
		TreeNode a = new TreeNode(5);
		TreeNode b = new TreeNode(4);
		TreeNode c = new TreeNode(7);
		TreeNode d = new TreeNode(12);
		root.left = a;
		a.left = b;
		a.right = c;
		root.right = d;
		ArrayList<ArrayList<Integer>> pathList = new Test25().FindPath(root, 22);
		for (ArrayList<Integer> path : pathList) {
			for (Integer value : path) {
				System.out.print(value + ", ");
			}
			System.out.println();
		}
	}
}
